### PR TITLE
decouple inference from Base

### PR DESCRIPTION
**Pull request by vtjnash**
having inference inside Base made it hard to modify and slow to compile the sysimg since bootstrapping made two complete passes over the code. additionally, it made inference vulnerable to sabotage -- intentional or unintentional -- by the users program.

to combat those issues, this creates a copy of the essential code required for running the inference into Core.Inference. this allows any part of the system (esp. Base) around inference to change, and even be recompiled or monkey-patched, without impacting the core inference algorithms.

@carnaval this also implies that you can build sysimg.jl without Core.Inference present, and that you can change the inference module in use at runtime.

inspired by @ihnorton's subversive activities in #11148

### CODE REVIEW

File: test/core.jl [hunk 1]
```diff
@@ -1164,7 +1164,7 @@ foo4129(a::Baz4129,b::Bar41291,args...) = foo4129(a,b.f,b,args...)
 foo4129(a::Baz4129,b::Bar41292,args...) = foo4129(a,b.f,b,args...)
 foo4129(a::Baz4129,args...)         = foo4129(a,a.b,args...)
 
-@test isa(foo4129(Baz4129(Bar41291(Foo4129())),1,2), Tuple{Baz4129,Bar4129,Foo4129,Int,Int})
+#@test isa(foo4129(Baz4129(Bar41291(Foo4129())),1,2), Tuple{Baz4129,Bar4129,Foo4129,Int,Int})
```
Comment (vtjnash) [inline test/core.jl]: @carnaval any idea why this change would have broken this test and reintroduced #4129?

**Comment by timholy** [comment]
:100: Will make it much easier to make disruptive changes, where getting inference working can be the main bottleneck.

I bet this speeds `make clean && make` quite a bit, as well as changes to files that aren't included in `coreimg.jl`.

**Comment by tkelman** [comment]
strange that julia-debug is segfaulting on the travis branch build but not the pr merge build?

File: test/core.jl (hunk 1 above)
Comment (vtjnash) [inline test/core.jl]: nevermind, i made a silly typo

**Comment by tkelman** [comment]
Some seemingly unrelated changes to inlining costs here?

Might be nice to not let `contrib/build_{sysimg,executable}.jl` bitrot too badly.

**Comment by vtjnash** [comment]
> Some seemingly unrelated changes to inlining costs here?

i removed support for floating point numbers and rewrote the algorithm in terms of integers. it should give the same answers however.

> Might be nice to not let contrib/build_{sysimg,executable}.jl bitrot too badly.

good point. in general, we might want to actually distribute the inference.ji file. although bootstrapping off an existing sys.ji file may also work.

**Comment by tkelman** [comment]
> i removed support for floating point numbers and rewrote the algorithm in terms of integers

Oh. Right. Not so unrelated then.

**Comment by tkelman** [comment]
Looks like this cuts off a good 2.5 minutes off of each AppVeyor build. Not too shabby.

**Comment by carnaval** [comment]
Had a quick look. The inference changes look reasonable (i.e. no semantic changes). I'm mostly afraid of all the potential places where it could be assumed that whatever the top module is, it is at least always the same. The possibilities of more flexible inference _and_ faster bootstrap are very enticing however.

**Comment by StefanKarpinski** [comment]
Is there any way to know about the top module assumptions other than merging and seeing what happens? Btw, I've been in favor of changing the top module from `Main` to a separate `Root` module for some time now (which is ironic since that's how Jeff originally had it and I convinced him to just use `Main` as the root).

File: src/codegen.cpp [hunk 2]
```diff
@@ -1282,27 +1282,7 @@ jl_value_t *jl_static_eval(jl_value_t *ex, void *ctx_, jl_module_t *mod,
             jl_value_t *f = jl_static_eval(jl_exprarg(e,0),ctx,mod,sp,ast,sparams,allow_alloc);
             if (f && jl_is_function(f)) {
                 jl_fptr_t fptr = ((jl_function_t*)f)->fptr;
-                if (fptr == &jl_apply_generic) {
```
Comment (Keno) [inline src/codegen.cpp]: What's up with this case getting removed?

Comment (vtjnash) [inline src/codegen.cpp]: this was support for a pre 0.1 deprecation from when ccall could only handle constant pointers. i removed jl_base_module, so i figured it was sensible to remove this special case also.

Comment (Keno) [inline src/codegen.cpp]: Ok, can this go on master to cut down the diff?

**Comment by vtjnash** [comment]
> I'm mostly afraid of all the potential places where it could be assumed that whatever the top module is, it is at least always the same

> Is there any way to know about the top module assumptions other than merging and seeing what happens? 

the top is a module that is assumed to provide certain functions required for basic operation. there is no documented list of these, but there are a few in inference.jl itself (to provide basic constant folding for some important operations) and a few in julia-parser (such as unsafe_cconvert). there is some inconsistency in that sometimes it is used to refer to functionality that is in Core, sometimes it refers to functionality that must be provided by Base for the base language spec to make sense, sometimes it refers to crucial functions that inference needs to optimize.

i think that this change actually doesn't need to alter that situation very much. it does however allow for the existence of multiple independent top modules, which hopefully this will make it clearer how to make the distinction between base and top in the future.

File: src/codegen.cpp (hunk 2 above)
Comment (vtjnash) [inline src/codegen.cpp]: Sure. Almost all of this diff is valid independent of moving inference itself so I could rearrange into many commits, or leave as one

Comment (vtjnash) [inline src/codegen.cpp]: do you have a preference, or can i just merge this?

Comment (Keno) [inline src/codegen.cpp]: I'd rather this be split out as a separate commit to keep commits atomic.

Comment (vtjnash) [inline src/codegen.cpp]: how's this now, better?

i think so anyways.

Comment (Keno) [inline src/codegen.cpp]: Yes, thanks.

Comment (tkelman) [inline src/codegen.cpp]: I quite like the granularity now, and the really descriptive commit messages are awesome too.

**Comment by vtjnash** [comment]
travis broken because of https://github.com/JuliaLang/julia/commit/7ec501d6#commitcomment-11228450

**Comment by tkelman** [comment]
Is `Warning: replacing module Inference` expected?

**Comment by vtjnash** [comment]
Yes. And it ooks like this (and #11343) are ready to merge.

**Comment by ihnorton** [comment]
+:100:

**Comment by tkelman** [comment]
Odd to have a warning during normal expected operation. That'll confuse people.

**Comment by ihnorton** [comment]
(took the liberty of rebasing the minor conflict, was at 164010f... merge finger is getting itchy)

**Comment by vtjnash** [comment]
CI is green, so: go for it?

**Comment by ihnorton** [comment]
YOLO

**Comment by sjkelly** [comment]
YOMO! (You Only Merge Once)

**Comment by ihnorton** [comment]
Cross-referencing #5155 as this makes the "modularize base in-place" discussion vastly more approachable.

(@sjkelly indeed :smile:)

**Comment by tkelman** [comment]
We should really get rid of that warning.

**Comment by ihnorton** [comment]
> We should really get rid of that warning.

@tkelman done (1b6f9be5fc7c547affa2a91a1852ff17cf7807fa)

**Comment by timholy** [comment]
I notice that when one is messing with files that are necessary for inference (e.g., `tuple.jl`), inference is always compiled twice. I'm not sure that was true formerly---if you didn't have to resort to `make clean` I think `make` would just compile things once. Is this more correct, or something that could be further optimized (safely)?

**Comment by tkelman** [comment]
Maybe if `inference0.ji` already exists, then that step could be made to skip the first stage build?

**Comment by vtjnash** [comment]
It's more correct, although rarely necessary (unless it impacts the results of inference). But I didn't altogether mind discouraging altering of those files, and even eventually shrinking them further :)
